### PR TITLE
Add a message to be logged when an email is sent successfully

### PR DESCRIPTION
**Pull request by lfdebrux**
We get a lot of user support queries asking whether an email was sent correctly or not; it would be nice to be able to always see a log message saying "email was sent". This PR adds a log message to the `DMNotifyClient.send_email_message()` method, so whatever logging there may be around the call you can always be sure at least one message was logged whether the email was sent successfully or not.

Oh and it does some refactoring as well.

### CODE REVIEW

File: dmutils/email/dm_notify.py [hunk 1]
```diff
@@ -9,6 +13,13 @@
 from dmutils.timing import logged_duration_for_external_request as log_external_request
 
 
+class DMNotifyEmail(NamedTuple):
+    to_email_address: str
+    template_name_or_id: str
+    personalisation: Optional[Dict[str, str]] = None
+    reference: Optional[str] = None
```
Comment (bjgill) [inline dmutils/email/dm_notify.py:20]: Nice.

Though, why is `reference` optional?

File: dmutils/email/dm_notify.py [hunk 2]
```diff
@@ -182,12 +198,14 @@ def send_email(
...
 
         except HTTPError as e:
-            self._log_email_error_message(to_email_address, template_name_or_id, reference, e)
+            self._log_email_error_message(email_obj, e)
             if isinstance(e.message, list) and \
                     any(msg["message"].startswith("Missing personalisation") for msg in e.message):
                 raise EmailTemplateError(str(e))
             raise EmailError(str(e))
 
+        self._log(logging.INFO, "Email sent to Notify successfully", email_obj)
```
Comment (bjgill) [inline dmutils/email/dm_notify.py:207]: What does this look like on the command line when being run as part of a script? Can we see the email object's fields as well as the message?

File: dmutils/email/dm_notify.py [hunk 3]
```diff
@@ -94,13 +105,13 @@ def get_reference(to_email_address, template_id, personalisation=None):
...
         :return: Hashed string 'reference' to be passed to client.send_email_notification or self.send_email
         """
-        personalisation_string = u','.join(
-            list(map(lambda x: u'{}'.format(x), personalisation.values()))
-        ) if personalisation else u''
-        details_string = u'|'.join([to_email_address, template_id, personalisation_string])
+        personalisation_string = ",".join(
+            list(map(lambda x: "{}".format(x), personalisation.values()))
+        ) if personalisation else ""
+        details_string = u"|".join([to_email_address, template_id, personalisation_string])
```
Comment (bjgill) [inline dmutils/email/dm_notify.py:111]:
```suggestion
        details_string = "|".join([to_email_address, template_id, personalisation_string])
```

**Review by bjgill — COMMENTED** [review]
A good improvement - this should save us a lot of repetitive logging in various scripts. Just a few questions.

File: dmutils/email/dm_notify.py (hunk 1 above)
Comment (lfdebrux) [reply dmutils/email/dm_notify.py:20]: Ah, that's a holdover from a different approach I tried and gave up on, I will fix that, thanks.

File: dmutils/email/dm_notify.py (hunk 3 above)
Comment (lfdebrux) [reply dmutils/email/dm_notify.py:111]: Thanks!

File: dmutils/email/dm_notify.py (hunk 2 above)
Comment (lfdebrux) [reply dmutils/email/dm_notify.py:207]: 🤔 I think not, but I'll get back to you on that

Comment (bjgill) [reply dmutils/email/dm_notify.py:207]: :+1: If not, might it be worth adding the `reference` to the message?

Comment (lfdebrux) [reply dmutils/email/dm_notify.py:207]: Yeah that's probably a good idea whatever the case; the handling of the `extra` field depends on the formatter and so could be different in different circumstances.

**Review by bjgill — APPROVED** [review]

**Pull request merged**